### PR TITLE
update baselines

### PR DESCRIPTION
**Pull request by frankxu2004**

### CODE REVIEW

**Comment by Milozms** [comment]
SDP-LSTM is from: Classifying relations via long short term memory networks along shortest dependency path, Xu Yan, Lili Mou, Ge Li, Yunchuan Chen, Hao Peng, and Zhi Jin. 2015. (http://www.aclweb.org/anthology/D15-1206)

**Comment by Milozms** [comment]
Please comment if there are differences between your implementation and my understanding as follows:

HypeNet: Improving Hypernymy Detection with an Integrated Path-based and Distributional Method, Vered Shwartz, Yoav Goldberg, Ido Dagan. (https://arxiv.org/pdf/1603.06076)

SDP-LSTM: Classifying relations via long short term memory networks along shortest dependency path, Xu Yan, Lili Mou, Ge Li, Yunchuan Chen, Hao Peng, and Zhi Jin. 2015. (http://www.aclweb.org/anthology/D15-1206)

CNN-PE: CNN with input word embedding + position embedding

BGRU+2ATT: Bidirectional GRU +word-level attention + sentence-level attention from Neural Relation Extraction with Selective Attention over Instances (http://www.aclweb.org/anthology/P16-1200)

Thanks!

**Comment by shanzhenren** [comment]
Looks good to me. @Frank please confirm. 

Regards,
Xiang Ren

**Comment by frankxu2004** [comment]
@Milozms 
For SDP-LSTM, it is a simplified version of the paper you mentioned. Please check https://nlp.stanford.edu/pubs/zhang2017tacred.pdf for the baseline implementation details, as well as CNN-PE. 
The BGRU-2ATT is also an improved version of the model in your paper mentioned.

**Comment by Milozms** [comment]
### BGRU+2ATT

Bidirectional GRU +word-level attention + sentence-level attention from Neural Relation Extraction with Selective Attention over Instances.

#### data_preprocess.py

Convert TACRED json to txt. (I think add this in readme is better)

#### initial.py：

- Init: 将数据集中的样例以entity pair作为key构建dict；然后按顺序，将entity pair相同的样例放到一起，entity pair相同的集合内部把label相同的放到一起。train_sen中每个元素为给定entity pair和label的所有句子，train_ans为相应的label。（entity pair和label都相同的记为一组）
- Separate: 将word embedding和position embedding分开，便于placeholder输入

#### train_GRU.py

- 将训练数据分为若干batch，每个batch的size为big_num（设定的参数），将batch输入train_step
- Train_step：将相同entity pair和label的样例（组）拆分为单个的样例，用total_shape记录每组的句子数（按顺序）

#### network.py

在计算sentence-level attention时，利用total_shape的信息，把同组（entity pair和label都相同）的样例放在一起计算，（batch_size是组size），每个组给出一个预测，之后计算loss。

**Comment by Milozms** [comment]
Hello @frankxu2004 , could you please provide the running instructions of "tacred-tf-ours" and "HypeNet" codes?  Thanks.
By the way, I think adding "data_preprocess.py" in the README of "Tensorflow-NRE" will be better since the original data doesn't include the txt-format TACRED data.

**Comment by frankxu2004** [comment]
Please check @Milozms . Sorry I cannot recall more instructions on this because many arguments are hardcoded in the code. It has been quite a while so it is messy.

**Comment by Milozms** [comment]
Okay. Thanks!

**Comment by shanzhenren** [comment]
@Milozms how's this thread going so far?

**Comment by Milozms** [comment]
I think there are no more problems with BGRU+2ATT. The others are still processing.

**Comment by Milozms** [comment]
Yes. It will be soon

**Comment by Milozms** [comment]
@frankxu2004 
I have problem running test_GRU.py. Every time it stops here and outputs nothing:
```
xiang@5455:~/maosen/TensorFlow-NRE$ CUDA_VISIBLE_DEVICES=0 python3 test_GRU.py TACRED 10900
/home/xiang/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
None index:  1
2018-07-02 17:01:17.260404: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-07-02 17:01:17.468234: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1344] Found device 0 with properties:
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:01:00.0
totalMemory: 10.91GiB freeMemory: 10.76GiB
2018-07-02 17:01:17.468269: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1423] Adding visible gpu devices: 0
2018-07-02 17:01:17.676459: I tensorflow/core/common_runtime/gpu/gpu_device.cc:911] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-07-02 17:01:17.676499: I tensorflow/core/common_runtime/gpu/gpu_device.cc:917]      0
2018-07-02 17:01:17.676504: I tensorflow/core/common_runtime/gpu/gpu_device.cc:930] 0:   N
2018-07-02 17:01:17.676733: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10413 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:01:00.0, compute capability: 6.1)
WARNING:tensorflow:From /home/xiang/anaconda3/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/datasets/base.py:198: retry (from tensorflow.contrib.learn.python.learn.datasets.base) is deprecated and will be removed in a future version.
Instructions for updating:
Use the retry module or similar alternatives.
WARNING:tensorflow:From /home/xiang/maosen/TensorFlow-NRE/network.py:132: softmax_cross_entropy_with_logits (from tensorflow.python.ops.nn_ops) is deprecated and will be removed in a future version.
Instructions for updating:

Future major versions of TensorFlow will allow gradients to flow
into the labels input on backprop by default.

See tf.nn.softmax_cross_entropy_with_logits_v2.
```

Another thing: The Tensorflow-NRE (BGRU+2ATT) directory contains a "table.png" which reports the P@N results. But it seems that test_GRU.py doesn't compute the P@N results.

**Comment by Milozms** [comment]
@shanzhenren 
There is a confusion on BGRU+2ATT model that it seems that we have inconsistent understanding on **”sentence-level attention”** (from http://www.aclweb.org/anthology/P16-1200). 
In my understanding, "sentence-level attention” (or sentence-selective attention) is only used in corpus-level RE setting, which is evaluated by **held-out** evaluation, which compares the relation predicted by the model with **facts in KB**, since the "sentence-level attention" attention binds all instances with the same entity pair and predict only one relation label for the entity pair. 
However, in sentence-level RE, the same entity pair may have different relations given different context. 
So I think using "sentence-level attention” in sentence-level RE is confusing. The experiment result also shows that BGRU+2ATT with sentence-level attention have a relatively lower performance (0.4855 F1 on TACRED, which is from the table on google docs). While I tried Bi-GRU with no attention and no additional information, and it gets 0.6008 F1 on TACRED.

**Comment by shanzhenren** [comment]
Right. Sentence-level attention does not work on sentence-level extraction (as one bag only has one relation mention). The attention should be removed.

**Comment by shanzhenren** [comment]
Maybe you could modify the BGRU code to adapt to our cases.

**Comment by Milozms** [comment]
I don't know whether this code is for sentence-level RE or corpus-level RE. 
If it is for sentence-level RE, most of the preprocessing are redundant, and I have a more simplified version. 
If it is for corpus-level RE, the evaluating code seems incomplete (see my former comment). And I think corpus-level RE doesn't need testing on TACRED dataset.

**Comment by Milozms** [comment]
From the perspective of implementation, I think separating the implementation of sentence-level and corpus-level is better. We can remain this for corpus-level, and my simplified implementation of sentence-level models is [here](https://github.com/shanzhenren/relation-extraction-analysis/tree/maosen/sentence-level-models/models). All RNN-like models can be integrated in this frame (except those using dependency paths). @cherry979988

**Comment by shanzhenren** [comment]
Yes, it might be clearer to separate these two sets of implementations. I think there is not too much changes on the model part; but probably significant changes on preprocessing and evaluation.

Xiang

**Comment by Milozms** [comment]
Preprocessing (on TACRED) and evaluation (for sentence-level) are already integrated in that framework and no more work on it is needed. Preprocessing for other dataset will be done soon.

**Comment by shanzhenren** [comment]
Great. I think the end goal is to have a clean and nice pipeline for both sentence and corpus-level extraction (and evaluation), using different models, on different (suitable) datasets.

Xiang

**Comment by shanzhenren** [comment]
But for this thread and repo; you just need to merge these baselines for the sentence-level extraction on TACRED dataset.

**Comment by Milozms** [comment]
So how to deal with the Tensorflow-NRE? Most of the codes are for corpus-level RE.

**Comment by shanzhenren** [comment]
Make it a special case for sentence level extraction where one instance is just one relation mention in a specific sentence.

Regards,
Xiang Ren

**Comment by shanzhenren** [comment]
BTW, is NYT’10 the only dataset used in THU’s NRE and BGRU-ATT work?

**Comment by frankxu2004** [comment]
@Milozms 
For the BGRU confusion, you could check my forked version of the original code. https://github.com/frankxu2004/Tensorflow-NRE The test setting here is to do corpus-level RE in correspondence to the NYT dataset they use, thus the sentence-level attention.
For my previous test result, it is just that each bag of instances has only one instance.

**Comment by Milozms** [comment]
@frankxu2004 What does "each bag of instances has only one instance" mean?
And what is the difference of your forked code and the original code? (except compatibility of Python 3 and Tensorflow 1.x)

**Comment by Milozms** [comment]
@frankxu2004 
The problem about running test_GRU.py is still unsolved. Every time it stops here and outputs nothing:
```
xiang@5455:~/maosen/TensorFlow-NRE$ CUDA_VISIBLE_DEVICES=0 python3 test_GRU.py TACRED 10900
/home/xiang/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
None index:  1
2018-07-02 17:01:17.260404: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-07-02 17:01:17.468234: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1344] Found device 0 with properties:
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:01:00.0
totalMemory: 10.91GiB freeMemory: 10.76GiB
2018-07-02 17:01:17.468269: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1423] Adding visible gpu devices: 0
2018-07-02 17:01:17.676459: I tensorflow/core/common_runtime/gpu/gpu_device.cc:911] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-07-02 17:01:17.676499: I tensorflow/core/common_runtime/gpu/gpu_device.cc:917]      0
2018-07-02 17:01:17.676504: I tensorflow/core/common_runtime/gpu/gpu_device.cc:930] 0:   N
2018-07-02 17:01:17.676733: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10413 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:01:00.0, compute capability: 6.1)
WARNING:tensorflow:From /home/xiang/anaconda3/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/datasets/base.py:198: retry (from tensorflow.contrib.learn.python.learn.datasets.base) is deprecated and will be removed in a future version.
Instructions for updating:
Use the retry module or similar alternatives.
WARNING:tensorflow:From /home/xiang/maosen/TensorFlow-NRE/network.py:132: softmax_cross_entropy_with_logits (from tensorflow.python.ops.nn_ops) is deprecated and will be removed in a future version.
Instructions for updating:

Future major versions of TensorFlow will allow gradients to flow
into the labels input on backprop by default.

See tf.nn.softmax_cross_entropy_with_logits_v2.
```

**Comment by Milozms** [comment]
@frankxu2004 what is the input format for tacred-tf-ours(SDP-LSTM) model? The input data of NYT is not provided.

**Comment by cherry979988** [comment]
Hi @frankxu2004 , I'm working on HypeNet and just want to confirm the running instructions with you:
1. Run preprocessing.py
2. Run shortest_dep.py twice, with 'FILE' hardcoded to be 'train' and 'test' separately
3. Run sdp.py
Not sure if I'm missing something. Thank you.

**Comment by shanzhenren** [comment]
@Milozms @cherry979988 let's first merge these baseline models (make sure they are runable and have sufficient instructions on README). Other stuffs can be done in the project repo.
